src/kv/mod.rs

```diff
@@ -0,0 +1,410 @@
+use std::collections::HashMap;
+use std::io::Write;
+
+use async_nats::jetstream::kv::Store;
+use async_nats::Client;
+use data_encoding::HEXUPPER;
+use ring::digest::{digest, SHA256};
+
+use crate::LinkDefinition;
+use crate::Result;
+
+mod cached;
+mod direct;
+
+pub use cached::*;
+pub use direct::*;
+
+const LINKDEF_PREFIX: &str = "LINKDEF_";
+const CLAIMS_PREFIX: &str = "CLAIMS_";
+const BUCKET_PREFIX: &str = "LATTICEDATA_";
+const SUBJECT_KEY: &str = "sub";
+
+#[async_trait::async_trait]
+pub trait KvStore {
+    /// Returns all links in the store
+    async fn get_links(&self) -> Result<Vec<LinkDefinition>>;
+
+    /// Returns all claims in the store
+    async fn get_claims(&self) -> Result<Vec<HashMap<String, String>>>;
+
+    /// Returns all provider claims in the store
+    async fn get_provider_claims(&self) -> Result<Vec<HashMap<String, String>>>;
+
+    /// Returns all actor claims in the store
+    async fn get_actor_claims(&self) -> Result<Vec<HashMap<String, String>>>;
+
+    /// Returns all links in the store that match the provided filter function. For some
+    /// implementations, this could be more efficient than fetching all links and filtering them in
+    /// memory.
+    async fn get_filtered_links<F>(&self, filter_fn: F) -> Result<Vec<LinkDefinition>>
+    where
+        F: FnMut(&LinkDefinition) -> bool + Send;
+
+    /// Returns the link definition for the given actor, link name, and contract ID, if it exists.
+    async fn get_link(
+        &self,
+        actor_id: &str,
+        link_name: &str,
+        contract_id: &str,
+    ) -> Result<Option<LinkDefinition>>;
+
+    /// Returns the claim for the given ID, if it exists.
+    async fn get_claim(&self, id: &str) -> Result<Option<HashMap<String, String>>>;
```

Comment:
I think `claims` is inherently pluralized — i.e. an actor or provider always has a _set_ of claims — so maybe this should be `get_claims` and the one above should be `get_all_claims` ?

Reply:
That is a good point, I'll update